src/test/arm/arch_timer.c

```diff
@@ -0,0 +1,62 @@
+#include "util.h"
+
+#include <sys/auxv.h>
+#include <sys/prctl.h>
+#include <stdio.h>
+
+long cntfrq(void) {
+  long c;
+  __asm__ __volatile__("mrs %0, cntfrq_el0" : "=r"(c));
+  return c;
+}
+
+long cntvct(void) {
+  long c;
+  __asm__ __volatile__("mrs %0, cntvct_el0" : "=r"(c));
+  return c;
+}
+
+long cntvctss(void) {
+  long c;
+  if (getauxval(AT_HWCAP2) & HWCAP2_ECV) {
+    __asm__ __volatile__(".arch armv8.6-a\nmrs %0, cntvctss_el0" : "=r"(c));
+  } else {
+    __asm__ __volatile__("mrs %0, cntvct_el0" : "=r"(c));
+  }
+  return c;
+}
+
+long initial_cntfrq;
+long initial_cntvct;
+
+void arch_timer_nops(void) {
+  __asm__ __volatile__("mrs xzr, cntfrq_el0");
+  __asm__ __volatile__("mrs xzr, cntvct_el0");
+  if (getauxval(AT_HWCAP2) & HWCAP2_ECV) {
+    __asm__ __volatile__("mrs xzr, cntvctss_el0");
+  }
+}
+
+void diversion_check(void) {
+  arch_timer_nops();
+  test_assert(initial_cntfrq == cntfrq());
+  test_assert(initial_cntvct < cntvct());
+  test_assert(initial_cntvct < cntvctss());
+  atomic_puts("diversion_check passed");
+}
+
+void breakpoint(void) {}
+
+int main(void) {
+  initial_cntfrq = cntfrq();
```

Comment:
I think this test should detect whether the kernel supports `PR_SET_TSC`.

One way to do that would be to give it a command-line option (something simple like "any arguments at all") that checks whether `PR_SET_TSC` is `EINVAL` and returns an exit code indicating that. Then we can call it (not under rr) from `arch_timer.run` and skip the test if that `PR_SET_TSC` is not supported.

Another way to do it would be in `record_syscall.cc` --- make `PR_SET_TSC` return `EINVAL` if the underlying kernel doesn't support it. Then the test can call `PR_SET_TSC` and bail out with `EXIT-SUCCESS` if it gets `EINVAL`. This would be good... currently when running under rr `PR_SET_TSC` claims to work even if it actually doesn't.

Reply:
Let's go with the first option. We've already had to accept that if `prctl(PR_SET_TSC)` isn't implemented in the kernel, things will be slightly broken, and `prctl(PR_SET_TSC)` not functioning correctly inside the recorded process is probably less of a problem than the nondeterministic timers.

Since we need to skip two tests (`arm/arch_timer` and `prctl_tsc`) , I decided to do the test as part of a separate program that is called by both tests' `.run` scripts.

In the process of doing so I discovered that I had added the new test to the wrong list so it wasn't actually running the `.run` script (which turned out to have a bug which meant that it would always fail), so I fixed that as well.